himbaechel/uarch/xilinx/pack_dsp_xc7.cc

```diff
@@ -0,0 +1,165 @@
+
+        PortRef& user = *ni->users.begin();
+        if (user.cell->type != id_DSP48E1_DSP48E1)
+            log_error("User %s of net %s is not a DSP block, but %s",
+                user.cell->name.c_str(ctx), ni->name.c_str(ctx), user.cell->type.c_str(ctx));
+    };
+
+    // see if any cascade outputs are connected
+    for (auto port : current_cell->ports) {
+        if (!boost::contains(port.first.str(ctx), "COUT")) continue;
+        NetInfo *cout_net = port.second.net;
+
+        if (cout_net == nullptr) continue;
+
+        check_illegal_fanout(cout_net, port.first.c_str(ctx));
+        PortRef& user = *cout_net->users.begin();
+        CellInfo *cout_cell = user.cell;
+        NPNR_ASSERT(cout_cell != nullptr);
+
+        if (cascaded_cell != nullptr && cout_cell != cascaded_cell)
+            log_error("the cascading outputs of DSP block %s are connected to different cells",
+                current_cell->name.c_str(ctx));
+
+        cascaded_cell = cout_cell;
+    }
+
+    if (cascaded_cell != nullptr) {
+        auto is_lower_bel = constr_z == BEL_LOWER_DSP;
+
+        cascaded_cell->cluster = root->name;
+        root->constr_children.push_back(cascaded_cell);
+        cascaded_cell->constr_x = 0;
+        // the connected cell has to be above the current cell,
+        // otherwise it cannot be routed, because the cascading ports
+        // are only connected to the DSP above
+        auto previous_y = (current_cell == root) ? 0 : current_cell->constr_y;
+        cascaded_cell->constr_y = previous_y + (is_lower_bel ? -5 : 0);
+        cascaded_cell->constr_z = constr_z;
+        cascaded_cell->constr_abs_z = true;
+
+        walk_dsp(root, cascaded_cell, is_lower_bel ? BEL_UPPER_DSP : BEL_LOWER_DSP);
+    }
+}
+
+void XC7Packer::pack_dsps()
+{
+    log_info("Packing DSPs..\n");
+
+    dict<IdString, XFormRule> dsp_rules;
+    dsp_rules[id_DSP48E1].new_type = id_DSP48E1_DSP48E1;
+    generic_xform(dsp_rules, true);
+
+    std::vector<CellInfo *> all_dsps;
+
+    for (auto &cell : ctx->cells) {
+        CellInfo *ci = cell.second.get();
+
+        auto add_const_pin = [&](PortInfo& port, std::string& pins, std::string& pin_name, std::string net) {
+            if (port.net && port.net->name == ctx->id(net)) {
+                ci->disconnectPort(port.name);
+                pins += " " + pin_name;
+            }
+        };
+
+        if (ci->type == id_DSP48E1_DSP48E1) {
+            all_dsps.push_back(ci);
+            auto gnd_attr = ctx->id("DSP_GND_PINS");
+            auto vcc_attr = ctx->id("DSP_VCC_PINS");
+
+            auto gnd_pins = str_or_default(ci->attrs, gnd_attr, "");
+            auto vcc_pins = str_or_default(ci->attrs, vcc_attr, "");
+
+            for (auto &port : ci->ports) {
+                std::string n = port.first.str(ctx);
+
+                if (boost::starts_with(n, "ACIN") || boost::starts_with(n, "BCIN") || boost::starts_with(n, "PCIN")) {
+                    if (port.second.net == nullptr)
```

Comment:
this should be changed to `port.second.net == nullptr || port.second.net->users.empty()` as we also want to skip the net if it's just dangling